saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/authentication/OpenSamlAuthenticationRequestFactory.java

```diff
@@ -58,6 +74,61 @@ public String createAuthenticationRequest(Saml2AuthenticationRequest request) {
 		);
 	}
 
+	@Override
+	public MultiValueMap<String, String> createAuthenticationRequestParameters(Saml2AuthenticationRequest request) {
```

Comment:
It surprises me a bit that the return type here isn't better defined, especially because of the name of the class. I'd expect a FooFactory to return instances of Foo, not take them as parameters. It also feels like the only option for further evolution will be to create additional interface methods that return still more sophisticated return types.

There is a somewhat similar class on the OAuth 2.0 side called `DefaultOAuth2AuthorizationRequestResolver`. It sits at a similar point in the authentication flow, and I wonder if introducing it at this point would help 1. make this API more similar to the OAuth support, 2. give the implementer more flexibility since they have the entire request available, and 3. future-proof the return type a bit better by having it return a `Saml2AuthenticationRequest`, like OAuth returns an `OAuth2AuthorizationRequest`.

Reply:
I agree that `Saml2AuthenticationRequest` would probably have been clearer if it was named `Saml2AuthenticationRequestConfiguration` and thus allowing `Saml2AuthenticationRequest` to be used as the result. 

One idea is to use  `Saml2AuthenticationRequestResult` or similar as a return type.

In the end, the goal is to have
```
Input: configuration data for an authentication request
Output: Form/Query Parameters to be used in the response
```

This interface is intentionally completely void of the calling stack (servlet vs reactive).
The filter can be smart enough to extract the request parameters needed for the implementation to do it's job